homeassistant/components/gardena_bluetooth/sensor.py

```diff
@@ -0,0 +1,125 @@
+"""Support for switch entities."""
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from datetime import datetime, timedelta, timezone
+
+from gardena_bluetooth.const import Battery, Valve
+from gardena_bluetooth.parse import Characteristic
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE, EntityCategory
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN
+from .coordinator import Coordinator, GardenaBluetoothEntity
+
+
+@dataclass
+class GardenaBluetoothSensorEntityDescription(SensorEntityDescription):
+    """Description of entity."""
+
+    char: Characteristic = field(default_factory=lambda: Characteristic(""))
+
+
+DESCRIPTIONS = (
+    GardenaBluetoothSensorEntityDescription(
+        key=Valve.activation_reason.uuid,
+        translation_key="activation_reason",
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        char=Valve.activation_reason,
+    ),
+    GardenaBluetoothSensorEntityDescription(
+        key=Battery.battery_level.uuid,
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.BATTERY,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        native_unit_of_measurement=PERCENTAGE,
+        char=Battery.battery_level,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Gardena Bluetooth sensor based on a config entry."""
+    coordinator: Coordinator = hass.data[DOMAIN][entry.entry_id]
+    entities: list[GardenaBluetoothEntity] = [
+        GardenaBluetoothSensor(coordinator, description)
+        for description in DESCRIPTIONS
+        if description.key in coordinator.characteristics
+    ]
+    entities.append(GardenaBluetoothRemainSensor(coordinator))
+    async_add_entities(entities)
+
+
+class GardenaBluetoothSensor(GardenaBluetoothEntity, SensorEntity):
+    """Representation of a sensor."""
+
+    entity_description: GardenaBluetoothSensorEntityDescription
+
+    def __init__(
+        self,
+        coordinator: Coordinator,
+        description: GardenaBluetoothSensorEntityDescription,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator, {description.key})
+        self._attr_native_value = None
+        self._attr_unique_id = f"{coordinator.address}-{description.key}"
```

Comment:
Might be nice to make a base class that takes care of constructing the unique id if you plan to keep using this pattern

Reply:
Will do a refactor separate, since it affect previous integrations.

Reply:
https://github.com/home-assistant/core/pull/96775